layer/contracts/runs.py

```diff
@@ -1,15 +1,227 @@
+        Includes single files and files in each resource directory.
+
+        :return: iterator of absolute file paths.
+        """
+        file_path = os.path.relpath(self.path)
+        if not os.path.exists(file_path):
+            raise FileNotFoundError(
+                f"resource file or directory: {self.path} in {os.getcwd()}"
+            )
+        if os.path.isfile(file_path):
+            yield file_path
+        if os.path.isdir(file_path):
+            for root, _, files in os.walk(file_path):
+                for f in files:
+                    dir_file_path = os.path.join(root, f)
+                    yield os.path.relpath(dir_file_path)
+
+
+FunctionDefinitionType = TypeVar("FunctionDefinitionType", bound="FunctionDefinition")
+
+
+class FunctionDefinition(abc.ABC):
+    def __init__(
+        self,
+        func: Any,
+        project_name: str,
+        account_name: Optional[str] = None,
+        version_id: Optional[uuid.UUID] = None,
+        description: str = "",
+        resource_paths: Optional[Set[ResourcePath]] = None,
+        language_version: Tuple[int, int, int] = _language_version(),
+    ) -> None:
+        self.func = func
+        self.version_id = version_id
+        self.project_name = project_name
+        self.account_name = account_name
+        self.resource_paths: Set[ResourcePath] = resource_paths or set()
+        self.description = description
+        self.language_version = language_version
+
+        layer_settings = func.layer
+        name = layer_settings.get_entity_name()
+        if name is None:
+            raise LayerClientException("Name cannot be empty")
+        self.name = name
+        fabric = layer_settings.get_fabric()
+        if fabric is None:
+            fabric = Fabric.F_LOCAL.value
+        self._fabric = fabric
+        self._dependencies = layer_settings.get_dependencies()
+        self.pip_packages = layer_settings.get_pip_packages()
+        self.pip_requirements_file = layer_settings.get_pip_requirements_file()
+
+        self.func_source = inspect.getsource(self.func)
+
+        self.source_code_digest = hashlib.sha256()
+        self.source_code_digest.update(self.func_source.encode("utf-8"))
+
+        self._pack()
+
+    def __repr__(self) -> str:
+        return f"FunctionDefinition({self.asset_type}, {self.name})"
+
+    @abc.abstractproperty
+    def asset_type(self) -> AssetType:
+        ...
+
+    @property
+    def asset_path(self) -> AssetPath:
+        return AssetPath(
+            asset_type=self.asset_type,
+            entity_name=self.name,
+            project_name=self.project_name,
+        )
+
+    @property
+    def dependencies(self) -> List[AssetPath]:
+        return [d.with_project_name(self.project_name) for d in self._dependencies]
+
+    @property
+    def entrypoint(self) -> str:
+        return f"{self.name}.pkl"
+
+    @property
+    def entity_path(self) -> Path:
```

Comment:
we should rename this, I think. It feels too ambiguous, especially given we also have `asset_path` too.

Reply:
I will do this in a separate PR for refactoring assets and entities.